programs/utilities/flower.sc

```diff
@@ -1,140 +1,296 @@
+  volume(pos1, pos2,
+    position = pos(_);
+    if(global_setting_ignore_biome || _is_flower_forest(position),
+      area_counts : _get_flower_id_at(position) += 1;
     );
   );
-  print('Filled ' + count + ' flowers');
+
+  p = player();
+  // print list of flowers
+  print(p, format('b Flowers in area: '));
+  for(area_counts,
+    print(p, str('%s: %d', global_flower_list : _i, _));
+  );
 );
 
-// fills area of given radius with flowers of given type
-filtered_flower_area(position, radius, flower_filter) -> (
-  // removes "minecraft:" in flower name if exists
-  fixed_flower_name = flower_filter - 'minecraft:';
-  flower_id = global_FLOWERS ~ fixed_flower_name;
+_run_count_area(pos1, pos2) -> (
+  // print pre-counting message
+  print(player(), str('Counting area of %d blocks...', __get_volume_size(pos1, pos2)));
 
-  // checks if given filter is a flower
-  if(flower_id == null,
-    print('not a valid flower: ' + flower_filter);
-    return();
-  );
+  // run counting task
+  task('count_area', pos1, pos2);
+);
 
+// fills an area. filter_flower_id must be null for no filter, or an index of global_flower_list
+fill_area(pos1, pos2, filter_flower_id) -> (
   count = 0;
-  without_updates(
-    scan(position, radius, 0, radius,
-      count += __filtered_flower_forest_fill(_, fixed_flower_name, flower_id);
-    );
+  volume(pos1, pos2,
+    // adds one for each block in volume that was placed
+    count += __place_flower_at(pos(_), filter_flower_id);
+  );
+
+  //print post-filling message
+  print(player(), str('Filled %d blocks', count));
+);
+
+_run_fill_area(pos1, pos2, flower_filter) -> (
+  p = player();
+
+  // check for OP
+  if(p ~ 'permission_level' < 2,
+    __error('You must be OP to fill blocks');
   );
-  print('Filled ' + count + ' flowers');
+  // check if glass_mode is not enabled and height of volume is more than 1
+  if(!global_setting_glass_mode && floor(pos1 : 1) != floor(pos2 : 1),
+    __error('You must have glass_mode setting enabled to fill an area with a height greater than 1');
+  );
+
+  // if flower is being filtered, get id of filter
+  if(flower_filter != null,
+    // removes "minecraft:" in name and gets index of flower
+    filter_flower_id = global_flower_list ~ (flower_filter - 'minecraft:');
+
+    // checks if given filter is a flower
+    if(filter_flower_id == null,
+      __error('Not a valid flower: ' + flower_filter);
+    ),
+  //else:
+    filter_flower_id = null;
+  );
+
+  // print pre-filling message
+  print(p, str('Filling area of %d blocks...', __get_volume_size(pos1, pos2)));
+
+  // run filling task
+  task('fill_area', pos1, pos2, filter_flower_id);
+);
+
+__print_flower_at(position) -> (
+  print(player(), str('Flower at %s: %s', join(', ', position), global_flower_list : _get_flower_id_at(position)));
 );
 
 _get_flower_id_at(position) -> (
-  x = position : 0;
-  z = position : 2;
+  x = floor(position : 0);
+  y = floor(position : 1);
+  z = floor(position : 2);
```

Comment:
you can do this as `[x, y, z] = position;`. Unpacking is nice.

Reply:
Didn't realize that was in scarpet, that's pretty helpful. Also, these coordinates need to be integers which is why this `floor()` is here, but now that I look at the rest of the code I don't think there's really a way floats could end up here, so I'll assume this is fine to change